### PR TITLE
RFC: support python 3.12 (but drop 3.8)

### PR DESCRIPTION
**Pull request by muxator**
This is an experiment aimed at supporting python 3.12, which is already at its second point release.

Formally, python 3.8 is still going to be supported for 6 months (https://devguide.python.org/versions). However, some packages have already dropped it.

In particular,  there is no scipy version supporting 3.8 and 3.12 at the same time, and this forces us to make a choice.

After bumping scipy to >= 1.11, poetry is able to resolve the dependencies.

At install time there were 2 more problems:

1. `numpy` did not install because python 3.12 dropped support for distutils (https://peps.python.org/pep-0632/). There are workarounds for this (https://numpy.org/devdocs/reference/distutils_status_migration.html), but bumping to 1.26 seems to be an official solution
2. `statsmodels` 0.13 did not compile under python 3.12, probably because it is compiled with `-Werror`, and there are some deprecation warnings when compiling its wheel. Bumping to 0.14 solved the issue.

This is just the easiest set of actions for allowing the project to run on python3.12. There might be ways to keep a more relaxed set of dependencies (for example, only install numpy 1.26 under python 3.12) if we want to be more permissive.

Local tests on my Fedora with python 3.12 completed fine. Opening the PR to see what the CI (and the fellow devs) think about it.

### CODE REVIEW

**Review by AldoGl — APPROVED** [review]

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -36,4 +36,4 @@ jobs:
       run: tox -e check-copyright
     - name: Misc checks
       run: |
-        tox -e bandit,safety
```
Comment (muxator) [inline .github/workflows/lint.yml:39]: If we no longer rely on safety we could also remove it from the dependencies in `pyproject.toml`.

Comment (AldoGl) [reply .github/workflows/lint.yml:39]: good idea!

Comment (muxator) [reply .github/workflows/lint.yml:39]: Never mind: Safety < 3.0.1 has grave issues (see https://github.com/pyupio/safety/issues/480) and the current version does not support performing a scan without logging into their service account (https://docs.safetycli.com/safety-docs/safety-cli-3/quick-start-guide):

> You will be unable to perform vulnerability scans unless you are authenticated. [Create an account and access your free trial here. ](https://platform.safetycli.com/register/)If you require assistance, please email [support@safetycli.com](mailto:support@safetycli.com).

Let's remove the dependency on Safety altogether.

**Comment by muxator** [comment]
This PR will remove the dependency on Safety (https://github.com/pyupio/safety), which has become a purely commercial product for all practical purposes.

Safety < 3.0.1 has grave issues (see https://github.com/pyupio/safety/issues/480) and the current version does not support performing a scan without logging into their service account (https://docs.safetycli.com/safety-docs/safety-cli-3/quick-start-guide):

> You will be unable to perform vulnerability scans unless you are authenticated. [Create an account and access your free trial here. ](https://platform.safetycli.com/register/)If you require assistance, please email [support@safetycli.com](mailto:support@safetycli.com).

Let's remove the dependency on Safety altogether.

File: tests/test_examples/test_main.py [hunk 2]
```diff
@@ -33,6 +35,13 @@ class TestMainExample(BaseMainExampleTestClass):
     script_path = EXAMPLE_MAIN_SCRIPT_PATH
     timeout = DEFAULT_SUBPROCESS_TIMEOUT
     nb_batches = 5
+    if sys.platform == "darwin":
+        # 2024-09-10: MethodOfMomentsLoss started giving slightly different
+        # values on MacOS when statsmodels was updated from 0.13 to 0.14. This
+        # captures the change, so that when a fix arrives upstream we can update
+        # the test again.
+        BEST_PARAMETERS_STR = "Best parameters found: [0.21 0.19 0.76]"
```
Comment (AldoGl) [inline tests/test_examples/test_main.py:43]: very nice solution for the moment

**Comment by AldoGl** [comment]
Thank you @muxator and @marcofavorito for the great help for this PR. As agreed, I will rebase and merge and we can tackle the OS compatibility issue another time

**Pull request merged**